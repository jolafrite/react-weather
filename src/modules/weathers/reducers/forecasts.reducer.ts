import { IWeatherForecast, IWeatherDailyForecast } from '../models';
import { ForecastApiActions } from "../actions";

export type State = {
  entries: {
    [id: number]: IWeatherDailyForecast
  };
};

export const initialState: State = {
  entries: {},
};

export const reducer = (
  state = initialState,
  action: ForecastApiActions.ForecastApiActionsUnion
): State => {
  const mergeEntries = (state: State, forecasts: IWeatherForecast[]): State => {
    const entities = forecasts.reduce((saved, forecast) => {
      return {
        ...saved,
        [forecast.id]: forecast.forecast
      };
    }, {});

    return {
      ...state,
      entries: {
        ...state.entries,
        ...entities,
      }
    };
  };

  switch(action.type) {
    case ForecastApiActions.LOAD_ONE_SUCCESS:
      return mergeEntries(state, [action.forecast]);

    default: {
      return state;
    }
  }
}
