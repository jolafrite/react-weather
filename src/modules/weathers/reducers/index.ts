import { combineReducers } from "redux";
import * as ForWeatherSearchPage from './weather-search-page.reducer';
import * as ForWeatherViewPage from './weather-view-page.reducer';
import * as ForWeathers from './weathers.reducer';

export type State = {
  readonly weathers: ForWeathers.State,
  readonly searchPage: ForWeatherSearchPage.State,
  readonly viewPage: ForWeatherViewPage.State,
};

export const reducer = combineReducers<State>({
  weathers: ForWeathers.reducer,
  searchPage: ForWeatherSearchPage.reducer,
  viewPage: ForWeatherViewPage.reducer,
});
