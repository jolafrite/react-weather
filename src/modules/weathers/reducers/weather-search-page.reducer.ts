import { WeatherSearchPageActions } from '../actions';

export type State = {
  status: string;
  error: Error | null;
  ids: number[];
  query: string | null;
};

export const initialState: State = {
  status: 'pending',
  error: null,
  ids: [],
  query: null
};

export const reducer = (
  state = initialState,
  action: WeatherSearchPageActions.WeatherSearchPageActionsUnion
): State => {
  switch(action.type) {

    case WeatherSearchPageActions.RESET: {
      return initialState;
    }

    case WeatherSearchPageActions.SEARCH: {
      return {
        ...state,
        status: 'loading',
        query: action.query
      };
    }

    case WeatherSearchPageActions.SEARCH_SUCCESS: {
      return {
        ...state,
        status: 'loaded',
        ids: action.weathers.map(weather => weather.city.id),
        error: null,
      };
    }

    case WeatherSearchPageActions.SEARCH_ERROR: {
      return {
        ...state,
        status: 'error',
        error: action.error,
      };
    }

    default: {
      return state;
    }
  };
};

export const getLoaded = (state: State) => state.status === 'loaded';
export const getLoading = (state: State) => state.status === 'loading' ;
export const getError = (state: State) => state.status === 'error';
export const getIds = (state: State) => state.ids;
export const getQuery = (state: State) => state.query;
