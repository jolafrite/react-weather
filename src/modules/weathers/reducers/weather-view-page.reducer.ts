import { WeatherViewPageActions } from '../actions';

export type State = {
  status: string;
  error: Error | null;
  id: number | null;
};

export const initialState: State = {
  status: 'pending',
  error: null,
  id: null,
};

export const reducer = (
  state = initialState,
  action: WeatherViewPageActions.WeatherViewPageActionsUnion
): State => {
  switch(action.type) {

    case WeatherViewPageActions.SELECT_ONE: {
      return {
        ...initialState,
        id: action.id
      };
    }

    case WeatherViewPageActions.LOAD_ONE: {
      return {
        ...state,
        status: 'loading',
      };
    }

    case WeatherViewPageActions.LOAD_ONE_SUCCESS: {
      return {
        ...state,
        status: 'loaded',
        id: action.weather.id,
        error: null,
      };
    }

    case WeatherViewPageActions.LOAD_ONE_ERROR: {
      return {
        ...state,
        status: 'error',
        error: action.error,
      };
    }

    default: {
      return state;
    }
  };
};

export const getLoaded = (state: State) => state.status === 'loaded';
export const getLoading = (state: State) => state.status === 'loading' ;
export const getError = (state: State) => state.status === 'error';
export const getId = (state: State) => state.id;
