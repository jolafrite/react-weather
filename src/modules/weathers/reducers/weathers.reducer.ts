import { IWeather } from "../models";
import { WeatherApiActions } from "../actions";

export type State = {
  entries: { [id: number]: IWeather};
};

export const initialState: State = {
  entries: {},
};

export const reducer = (
  state = initialState,
  action: WeatherApiActions.WeatherApiActionsUnion
): State => {
  const mergeEntries = (state: State, weathers: IWeather[]): State => {
    const entities = weathers.reduce((saved, weather) => {
      return {
        ...saved,
        [weather.city.id]: weather
      };
    }, {});

    return {
      ...state,
      entries: {
        ...state.entries,
        ...entities,
      }
    };
  }

  switch(action.type) {

    case WeatherApiActions.LOAD_LIST_SUCCESS:
      return mergeEntries(state, action.weathers);

    case WeatherApiActions.LOAD_ONE_SUCCESS:
      return mergeEntries(state, [action.weather]);

    default: {
      return state;
    }
  }
}
