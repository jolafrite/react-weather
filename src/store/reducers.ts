import { combineReducers } from "redux";
import { connectRouter, RouterState } from 'connected-react-router';
import { History } from "history";
import * as FromWeathers from '../modules/weathers/reducers';

export type RootState = {
  readonly router: RouterState,
  readonly weathers: FromWeathers.State,
};

export const createRootReducer = (history: History) => combineReducers({
  router: connectRouter(history),
  weathers: FromWeathers.reducer,
});
