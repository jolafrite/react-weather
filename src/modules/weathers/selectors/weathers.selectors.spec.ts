import {
  getWeatherEntitiesState,
  getWeatherEntities
} from "./weathers.selectors";
import { getWeathersState } from "../reducers";

describe('selectors: weather', () => {

  describe('getWeatherState', () => {
    test('should return the root state sub state weathers ', () => {
      const state = {
        router: {
          router: 'router state'
        },
        weathers: {
          weathers: 'weathers state'
        }
      };

      expect(
        getWeathersState(state as any)
      ).toMatchSnapshot();
    });
  });

  describe('getWeatherEntitiesState', () => {
    test('should return the weather entries state', () => {
      const state = {
        weathers: {
          'weathers': 'weathers state'
        }
      };

      expect(
        getWeatherEntitiesState(state as any)
      ).toMatchSnapshot();
    });
  });

  describe('getWeatherEntities', () => {
    test('should return the entries value', () => {
      const state = {
        weathers: {
          weathers: {
            entries: {
              1: { id: 1 },
              2: { id: 2 },
            }
          }
        }
      };

      expect(
        getWeatherEntities(state as any)
      ).toMatchSnapshot();
    });
  });

});
