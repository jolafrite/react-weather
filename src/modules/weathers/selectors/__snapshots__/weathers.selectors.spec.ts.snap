// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors: weather getWeatherEntities should return the entries value 1`] = `
Object {
  "1": Object {
    "id": 1,
  },
  "2": Object {
    "id": 2,
  },
}
`;

exports[`selectors: weather getWeatherEntitiesState should return the weather entries state 1`] = `"weathers state"`;

exports[`selectors: weather getWeatherState should return the root state sub state weathers  1`] = `
Object {
  "weathers": "weathers state",
}
`;
